docs/development/design/fsm.md

```diff
@@ -0,0 +1,57 @@
+# Nexd State Machine
+
+> [Issue #885](https://github.com/nexodus-io/nexodus/issues/885)
+
+## Summary
+
+State management in nexd is becoming more complicated.
+We need a mechanism to allow us to simplify that logic and not rely on crash/restart to retrigger things.
+
+## Proposal
+
+I propose that we implement a Finite State Machine inside of nexd that operates as follows:
+
+```mermaid
+---
+title: Nexd State Machine
+---
+stateDiagram-v2
+    [*] --> IDLE
+    IDLE --> AUTH: "login"
+    AUTH --> READY: "login_success"
+    AUTH --> IDLE: "login_failed"
+    READY --> PEERING: "start"
+    READY --> IDLE: "logout"
+    PEERING --> READY: "stop"
+    PEERING --> READY: "peering_error"
+    PEERING --> IDLE: "authentication_error"
+    PEERING --> IDLE: "logout"
+```
+
```

Comment:
I think this FSM is assuming that all the states will be forced through nexctl. I think we need to consider UX actions as well in this FSM, until and unless we assumes that UX is just real only. 

Just focusing on the device end-to-end workflow, i think we can take two possible action from UX
1) Delete device --> Map it to nexd logout
2) Block device  (doesn't exist yet) --> Keep the device mark in apiserver db, invalidate the token --> `nexd logout` and move to idle state. Further attempt by user to onboard the same device will be rejected.

Reply:
> I think this FSM is assuming that all the states will be forced through nexctl.

It is not. "authentication_error" or and "peering_error" are pushed by next in response to API errors.
401 = authentication error, and 5XX but not the slow down error code would be "peering_error".

If we had bi-dir RPC we'd have new transitions for "deleted" and "blocked" which would put trigger the FSM to move to the appropriate state.

Reply:
I think we need to define the parallel fsm for control plane as well. Device related changes that are triggered through UX actions by can be pushed to nexd as a device state changes, that nexd can listen to and update it's local FSM.

Reply:
The control plane does not need a fsm. CRUD with validations should be enough for the API.